File: packages/cli/src/daemon/sensors/railway-health.ts
import type { Sensor, SensorContext } from "./index.js";
import type { SensorResult } from "../../types/daemon.js";

export class RailwayHealthSensor implements Sensor {
	id: string;
	name: string;
	private serviceKey: "symphonyUrl" | "arcanUrl" | "lagoUrl" | "autonomicUrl";

	constructor(
		serviceId: string,
		serviceName: string,
		serviceKey: "symphonyUrl" | "arcanUrl" | "lagoUrl" | "autonomicUrl",
	) {
		this.id = `railway-${serviceId}`;
		this.name = `Railway ${serviceName}`;
		this.serviceKey = serviceKey;
	}

	async run(ctx: SensorContext): Promise<SensorResult> {
		const timestamp = new Date().toISOString();
		const baseUrl = ctx.config[this.serviceKey];

		if (!baseUrl) {
			return {
				sensorId: this.id,
				status: "unknown",
				message: `${this.name}: URL not configured (set daemon.${this.serviceKey})`,
				timestamp,
			};
		}

		const healthPaths = ["/healthz", "/health"];
		const results: Record<string, unknown> = {};

		for (const path of healthPaths) {
			const url = `${baseUrl.replace(/\/$/, "")}${path}`;
			const start = Date.now();
			try {
				const headers: Record<string, string> = {};
				if (ctx.token) {
					headers.Authorization = `Bearer ${ctx.token}`;
				}
				const res = await fetch(url, {
					headers,
					signal: AbortSignal.timeout(10_000),
				});
				const latency = Date.now() - start;

				if (res.ok) {
					results.health = {
						path,
						status: res.status,
						latencyMs: latency,
						ok: true,
					};

					// For Symphony, also try to get state
					if (this.serviceKey === "symphonyUrl") {
						await this.checkSymphonyState(baseUrl, ctx.token, results);
					}

					return {
						sensorId: this.id,
						status: "healthy",
						message: `${this.name} responding on ${path}`,
						latencyMs: latency,
						timestamp,
						data: results,
					};
				}

				results[path] = { status: res.status, latencyMs: latency, ok: false };
			} catch (err) {
				const latency = Date.now() - start;
				results[path] = {
					status: 0,
					latencyMs: latency,
					ok: false,
					error: err instanceof Error ? err.message : String(err),
				};
			}
		}

		return {
			sensorId: this.id,
			status: "down",
			message: `${this.name} unreachable`,
			timestamp,
			data: results,
		};
	}

	private async checkSymphonyState(
		baseUrl: string,
		token: string | null,
		results: Record<string, unknown>,
	): Promise<void> {
		try {
			const headers: Record<string, string> = {};
			if (token) {
				headers.Authorization = `Bearer ${token}`;
			}
			const res = await fetch(
				`${baseUrl.replace(/\/$/, "")}/api/v1/state`,
				{
					headers,
					signal: AbortSignal.timeout(10_000),
				},
			);
			if (res.ok) {
				const state = await res.json();
				results.symphonyState = state;
			}
		} catch {
			// Symphony state check is best-effort
		}
	}
}

export function createRailwaySensors(): RailwayHealthSensor[] {
	return [
		new RailwayHealthSensor("symphony", "Symphony", "symphonyUrl"),
		new RailwayHealthSensor("arcan", "Arcan", "arcanUrl"),
		new RailwayHealthSensor("lago", "Lago", "lagoUrl"),
		new RailwayHealthSensor("autonomic", "Autonomic", "autonomicUrl"),
	];
}

File: crates/broomva-cli/src/daemon/sensors/railway_health.rs
use std::pin::Pin;
use std::time::Instant;

use super::{Sensor, SensorContext, SensorResult, SensorStatus};

/// Generic Railway service sensor — checks Symphony, Arcan, Lago, Autonomic endpoints.
pub struct RailwayHealthSensor;

impl Sensor for RailwayHealthSensor {
    fn id(&self) -> &str {
        "railway_health"
    }

    fn run<'a>(
        &'a self,
        ctx: &'a SensorContext,
    ) -> Pin<Box<dyn std::future::Future<Output = SensorResult> + Send + 'a>> {
        Box::pin(async move {
            let start = Instant::now();

            let services: Vec<(&str, Option<&str>)> = vec![
                ("symphony", ctx.symphony_url.as_deref()),
                ("arcan", ctx.arcan_url.as_deref()),
                ("lago", ctx.lago_url.as_deref()),
                ("autonomic", ctx.autonomic_url.as_deref()),
            ];

            let mut results = serde_json::Map::new();
            let mut configured = 0u32;
            let mut healthy = 0u32;

            for (name, url) in &services {
                if let Some(url) = url {
                    configured += 1;
                    let health_url = format!("{url}/healthz");
                    let ok = ctx
                        .client
                        .get(&health_url)
                        .send()
                        .await
                        .map(|r| r.status().is_success())
                        .unwrap_or(false);
                    if ok {
                        healthy += 1;
                    }
                    results.insert(
                        name.to_string(),
                        serde_json::json!({ "url": url, "healthy": ok }),
                    );
                } else {
                    results.insert(name.to_string(), serde_json::json!({ "configured": false }));
                }
            }

            let latency = start.elapsed().as_millis() as u64;

            let (status, message) = if configured == 0 {
                (
                    SensorStatus::Healthy,
                    "no railway services configured".into(),
                )
            } else if healthy == configured {
                (
                    SensorStatus::Healthy,
                    format!("{healthy}/{configured} services healthy"),
                )
            } else if healthy > 0 {
                (
                    SensorStatus::Degraded,
                    format!("{healthy}/{configured} services healthy"),
                )
            } else {
                (
                    SensorStatus::Down,
                    format!("0/{configured} services healthy"),
                )
            };

            SensorResult {
                sensor_id: self.id().into(),
                status,
                message,
                latency_ms: Some(latency),
                details: Some(serde_json::Value::Object(results)),
            }
        })
    }
}
